controllers/openstackmachine_controller.go

```diff
@@ -292,11 +302,155 @@ func (r *OpenStackMachineReconciler) reconcileDelete(scope scope.Scope, cluster
+							Kind:       openStackMachine.Kind,
+							Name:       openStackMachine.Name,
+							UID:        openStackMachine.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressClaimSpec{
+					PoolRef: *poolRef,
+				},
+			}
+			if err := r.Client.Create(ctx, claim); err != nil {
+				conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "Failed to reconcile floating IP claims: %v", err)
+				return nil, err
+			}
+			scope.Logger().Info("Created IPAddressClaim", "name", claim.Name)
+		} else {
+			conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "Failed to reconcile floating IP claims: %v", err)
+			return nil, err
+		}
+	}
+
+	if claim.Status.AddressRef.Name == "" {
+		scope.Logger().Info("Waiting for IPAddressClaim to be allocated", "name", claim.Name)
+		openStackMachine.Status.FloatingAddressFromPoolReady = pointer.Bool(false)
+		conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolWaitingForIpamProviderReason, clusterv1.ConditionSeverityWarning, "")
+		return claim, nil
+	}
+	conditions.MarkTrue(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition)
+	return claim, nil
+}
+
+func (r *OpenStackMachineReconciler) associateIPAddressFromIPAddressClaim(ctx context.Context, scope scope.Scope, openStackMachine *infrav1.OpenStackMachine, openStackCluster *infrav1.OpenStackCluster, instanceStatus *compute.InstanceStatus, instanceNS *compute.InstanceNetworkStatus, claim *ipamv1.IPAddressClaim) error {
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return err
+	}
+	computeService, err := compute.NewService(scope)
+	if err != nil {
+		return err
+	}
+
+	address := &ipamv1.IPAddress{}
+	addressKey := client.ObjectKey{Namespace: openStackMachine.Namespace, Name: claim.Status.AddressRef.Name}
+
+	if err := r.Client.Get(ctx, addressKey, address); err != nil {
+		return err
+	}
+
+	instanceAddresses := instanceNS.Addresses()
+	for _, instanceAddress := range instanceAddresses {
+		if instanceAddress.Address == address.Spec.Address {
+			openStackMachine.Status.FloatingAddressFromPoolReady = pointer.Bool(true)
+			conditions.MarkTrue(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition)
+			return nil
+		}
+	}
+	fip, err := networkingService.GetFloatingIP(address.Spec.Address)
+	if err != nil {
+		return err
+	}
+	if fip == nil {
+		conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "floating IP does not exist")
+		return fmt.Errorf("floating IP %q does not exist", address.Spec.Address)
+	}
+
+	port, err := computeService.GetManagementPort(openStackCluster, instanceStatus)
```

Comment:
I hate this function, and I wrote it. Everything about it is nasty. It contains so many assumptions, for example that the machine has an interface on the default cluster network. What if this is a spine leaf deployment and it doesn't?

Just spitballing: how about instead of defining a floating IP against the machine, we defined it against a specific port? What would the UX of that be like?

Alternatively, can we place constraints on the floating ip pool used? E.g. it *must* allocate a floating IP from the one and only `ExternalNetwork`, and therefore we need to pick a port on a subnet which is attached to the default router?

For me this is the biggest risk in this API change. I would really like to understand this.